### ISSUE TITLE
cmus on ARM

### ISSUE THREAD

**Issue by marosd** [cmus/cmus#229]
Hi everyone,

At first I would like to say hello to the community. I have a noob question about cross compilation of cmus on ARM where I have installed Arago linux. I am realy new in Linux so sorry if my questions are stupid.

I downloaded cmus source codes and I tried compile cmus on my Ubuntu and it works well. I only installed ALSA plugin and run configure script. The script shows me some dependances of cmus. I run make commnad -> sudo make install. After that I can run cmus and play some wav files. (flac and mp3 not because I did not installed any libraries). Here is my first question:

1. How could I add these libraries into cmus source codes without installation of these libraries? I mean, when I run this command:
sudo apt-get build-dep cmus
so it will automatically install dependance libraries into my filesystem (somewhere), right? Then cmus finds this libraries and everything is OK, compile is OK. But my question is, if there is some other way - without instalation of libraries into my filesystem. For example I will download the FLAC library-> tar -xvf FLAC -> and I will simply copy it into general cmus folder with source codes. Then I run configure command and make commnad and thats it...this way not work for me. But if this way somehow works it will be great because cross compilation will be easier, I think (?).

2. If not, so how could I cross compile cmus? I have to cross-compile all dependance libraries separately -> then cross compile cmus -> and install on my target system?

3. If 2nd question was good idea, could you give me some guidance? Some step by step manual? I am really new in Linux and this stuff and i am lost :(

When I was trying to cross-compile MPlayer everything works because all libraries were in the mplayer folder. So I simple export PATH where I have ARM toolchain and run this:
./configure --enable-cross-compile --cc=$(TOOL_CHAIN_PREFIX)-gcc --host-cc=gcc --as=$(TOOL_CHAIN_PREFIX)--as --ar=$(TOOL_CHAIN_PREFIX)--ar --ranlib=$(TOOL_CHAIN_PREFIX)--ranlib --target=arm-linux 

And everything was done. But I like CMUS player, so I want it on my target instead of mplayer :)

Thank you for help!!!

Maros

**Comment by flyingmutant** [comment]
Sorry, I have no solid guide for you. They only thing I can say is that cmus has no features which will make its cross-compilation harder than any other unix program with dependencies on external dynamic libraries.

**Comment by marosd** [comment]
Hi,

But I am a new in Linux, so everything was hard for me - a lot of things and no time. But after 4 days I was able to successful cross-compile CMUS on my ARM device :) 

I am testing CMUS right know and WAV files are playing good, but I have a problem with FLAC files - after 3 minutes sound is distorted...

**Comment by marosd** [comment]
Another problem is that I cannot play mp3 files. Here is ip folder:
TILINUX/BUILDCODES/CMUSARM/lib/cmus/ip$ ls
aac.so  flac.so  mad.so  vorbis.so  wav.so

after CMUS build and run ./cmus --p:

Input Plugins: /home/maros/TILINUX/BUILDCODES/CMUSARM/lib/cmus/ip
  flac:
    Priority: 50
    File Types: flac fla
    MIME Types:
  vorbis:
    Priority: 50
    File Types: ogg oga ogx
    MIME Types: application/ogg audio/x-ogg
  wav:
    Priority: 50
    File Types: wav
    MIME Types:
  aac:
    Priority: 50
    File Types: aac
    MIME Types: audio/aac audio/aacp

Output Plugins: /home/maros/TILINUX/BUILDCODES/CMUSARM/lib/cmus/op
  alsa

Do you know where is the problem? Thanks

**Comment by marosd** [comment]
Ok, I have made some research.

I tried to compile ffmpeg.c file from cmus directory manually like this:
my/toolchain/arm-linux-gnueabihf-gcc ffmpeg.c -o ffmpeg

and I got error that it cannot recognize libraries..

then I realized that I forgot to add search path for libraries. So I added -I path exactly same as ./configure script:

my/toolchain/arm-linux-gnueabihf-gcc ffmpeg.c -o ffmpeg -I/home/maros/TILINUX/BUILDCODES/CMUSARM/include

NOW, Libraries are recognized but it yells on me:

ffmpeg.c: In function ‘ffmpeg_open’:
ffmpeg.c:238:4: warning: ‘codec_name’ is deprecated (declared at /home/maros/TILINUX/BUILDCODES/CMUSARM/include/libavcodec/avcodec.h:1244) [-Wdeprecated-declarations]
ffmpeg.c:251:4: warning: ‘codec_name’ is deprecated (declared at /home/maros/TILINUX/BUILDCODES/CMUSARM/include/libavcodec/avcodec.h:1244) [-Wdeprecated-declarations]
ffmpeg.c: In function ‘ffmpeg_fill_buffer’:
ffmpeg.c:350:2: warning: ‘avcodec_alloc_frame’ is deprecated (declared at /home/maros/TILINUX/BUILDCODES/CMUSARM/include/libavcodec/avcodec.h:3612) [-Wdeprecated-declarations]
ffmpeg.c:366:5: warning: ‘avcodec_free_frame’ is deprecated (declared at /home/maros/TILINUX/BUILDCODES/CMUSARM/include/libavcodec/avcodec.h:3637) [-Wdeprecated-declarations]
ffmpeg.c:429:4: warning: ‘avcodec_free_frame’ is deprecated (declared at /home/maros/TILINUX/BUILDCODES/CMUSARM/include/libavcodec/avcodec.h:3637) [-Wdeprecated-declarations]
/home/maros/TILINUX/ti-sdk-am335x-evm-07.00.00.00/linux-devkit/sysroots/i686-arago-linux/usr/bin/../arm-linux-gnueabihf/libc/usr/lib/arm-linux-gnueabihf/4.7.3/crt1.o: In function `_start':
(.text+0x30): undefined reference to `main'
/tmp/ccfUXB3m.o: In function `xmalloc':
ffmpeg.c:(.text+0x22): undefined reference to `malloc_fail'
/tmp/ccfUXB3m.o: In function `ffmpeg_input_create':
ffmpeg.c:(.text+0xce): undefined reference to `av_new_packet'
/tmp/ccfUXB3m.o: In function `ffmpeg_input_free':
ffmpeg.c:(.text+0x108): undefined reference to `av_free_packet'
/tmp/ccfUXB3m.o: In function `ffmpeg_init':
ffmpeg.c:(.text+0x1ea): undefined reference to `av_log_set_level'
ffmpeg.c:(.text+0x1ee): undefined reference to `av_register_all'
/tmp/ccfUXB3m.o: In function `ffmpeg_open':
ffmpeg.c:(.text+0x236): undefined reference to `avformat_open_input'
ffmpeg.c:(.text+0x254): undefined reference to `__debug_print'
ffmpeg.c:(.text+0x264): undefined reference to `avformat_find_stream_info'
ffmpeg.c:(.text+0x282): undefined reference to `__debug_print'
ffmpeg.c:(.text+0x2da): undefined reference to `__debug_print'
ffmpeg.c:(.text+0x2ec): undefined reference to `avcodec_find_decoder'
ffmpeg.c:(.text+0x310): undefined reference to `__debug_print'
ffmpeg.c:(.text+0x33a): undefined reference to `avcodec_open2'
ffmpeg.c:(.text+0x35c): undefined reference to `__debug_print'
ffmpeg.c:(.text+0x374): undefined reference to `avformat_close_input'
ffmpeg.c:(.text+0x3b2): undefined reference to `avcodec_close'
ffmpeg.c:(.text+0x3bc): undefined reference to `avformat_close_input'
ffmpeg.c:(.text+0x3d6): undefined reference to `swr_alloc'
ffmpeg.c:(.text+0x3e4): undefined reference to `av_get_default_channel_layout'
ffmpeg.c:(.text+0x3fa): undefined reference to `av_opt_set_int'
ffmpeg.c:(.text+0x406): undefined reference to `av_get_default_channel_layout'
ffmpeg.c:(.text+0x41c): undefined reference to `av_opt_set_int'
ffmpeg.c:(.text+0x438): undefined reference to `av_opt_set_int'
ffmpeg.c:(.text+0x454): undefined reference to `av_opt_set_int'
ffmpeg.c:(.text+0x46c): undefined reference to `av_opt_set_sample_fmt'
ffmpeg.c:(.text+0x4c6): undefined reference to `av_opt_set_sample_fmt'
ffmpeg.c:(.text+0x4e6): undefined reference to `av_opt_set_sample_fmt'
ffmpeg.c:(.text+0x506): undefined reference to `av_opt_set_sample_fmt'
ffmpeg.c:(.text+0x50e): undefined reference to `swr_init'
ffmpeg.c:(.text+0x532): undefined reference to `channel_map_init_waveex'
/tmp/ccfUXB3m.o: In function `ffmpeg_close':
ffmpeg.c:(.text+0x55a): undefined reference to `avcodec_close'
ffmpeg.c:(.text+0x564): undefined reference to `avformat_close_input'
ffmpeg.c:(.text+0x56e): undefined reference to `swr_free'
/tmp/ccfUXB3m.o: In function `ffmpeg_fill_buffer':
ffmpeg.c:(.text+0x5ae): undefined reference to `avcodec_alloc_frame'
ffmpeg.c:(.text+0x5c6): undefined reference to `av_free_packet'
ffmpeg.c:(.text+0x5d0): undefined reference to `av_read_frame'
ffmpeg.c:(.text+0x5e0): undefined reference to `avcodec_free_frame'
ffmpeg.c:(.text+0x622): undefined reference to `av_new_packet'
ffmpeg.c:(.text+0x646): undefined reference to `avcodec_decode_audio4'
ffmpeg.c:(.text+0x652): undefined reference to `av_free_packet'
ffmpeg.c:(.text+0x694): undefined reference to `swr_convert'
ffmpeg.c:(.text+0x6c6): undefined reference to `avcodec_free_frame'
/tmp/ccfUXB3m.o: In function `ffmpeg_seek':
ffmpeg.c:(.text+0x7d2): undefined reference to `av_rescale_q'
ffmpeg.c:(.text+0x7e0): undefined reference to `avcodec_flush_buffers'
ffmpeg.c:(.text+0x800): undefined reference to `av_seek_frame'
/tmp/ccfUXB3m.o: In function `ffmpeg_read_comments':
ffmpeg.c:(.text+0x87c): undefined reference to `comments_add_const'
ffmpeg.c:(.text+0x894): undefined reference to `av_dict_get'
ffmpeg.c:(.text+0x8a6): undefined reference to `keyvals_terminate'
/tmp/ccfUXB3m.o: In function `ffmpeg_codec_profile':
ffmpeg.c:(.text+0xa2c): undefined reference to `av_get_profile_name'
collect2: error: ld returned 1 exit status

**Comment by flyingmutant** [comment]
These are linker errors — they are due to the fact that `ffmpeg.c` depends on other functions in cmus (and ffmpeg), and you do not link those together.

**Comment by marosd** [comment]
OK, but ./configure script shows me this include paths...so, it would be great if you give me right command how to compile ffmpeg.c from cmus directory with gcc compiler.

OR this is my command for cross-compile:
PKG_CONFIG_PATH="/home/maros/TILINUX/BUILDCODES/CMUSARM/lib/pkgconfig" ./configure prefix=/home/maros/TILINUX/BUILDCODES/CMUS CROSS=/home/blaaaaaah/bin/arm-linux-gnueabihf-

this is from terminal:
checking for FFMPEG_LIBS (pkg-config)... -pthread -L/home/maros/TILINUX/BUILDCODES/CMUSARM/lib -lavformat -lavcodec -lasound -lbz2 -lz -lswresample -lavutil -lrt -lm  
checking for FFMPEG_CFLAGS (pkg-config)... -I/home/maros/TILINUX/BUILDCODES/CMUSARM/include  
checking for header <libavcodec/avcodec.h>... yes
checking for successful build of ffmpeg.c... no

So it finds all dependances, isnt it?

could you please help me?

**Comment by flyingmutant** [comment]
I guess you need to learn how C programs are built first :-)

There are two distinct steps: compilation and linking. During compilation, `.c` files (together with all `#include`s) are translated to `.o` files. During linking, `.o` files from your program (and libraries) are combined together to produce the final binary.

You do not need to compile `ffmpeg.c` separately from the rest of cmus. Configure and cross-compile cmus as a whole, it will automatically build all plugins for which you have the dependencies.

FFMPEG is a bit special from other plugins, though — because ffmpeg/libav have 100500 incompatible API versions, cmus' plugin can possibly fail to compile with the one you have installed (this is checked during the configure step, `checking for successful build of ffmpeg.c...`).

**Comment by marosd** [comment]
I think, you dont understand me. 

I am trying to build CMUS on my platform - but without success. When I build CMUS for UBUNTU everything is working and I am able to play everything what I want. When I build CMUS for UBUNTU and run ./cmus --p command I see all input plugins and FFMPEG too. In UBUNTU is ffmpeg.c builded OK: checking for successful build of ffmpeg.c ... yes

When I am building CMUS for ARM I don see any FFMEG input plugins and I am not able to play .wma files for example. And I get checking for successful build of ffmpeg.c ... no

I want to solve this and I asking for help I think on the right place - here. I compile separately ffmpeg.c source code because I do not know why ./configure script fails on this.

**Comment by flyingmutant** [comment]
Can you test if `pkg-config` is working for ffmpeg in your ARM toolchain?

Here is a snippet from configure file which tries to compile ffmpeg:
```sh
check_ffmpeg()
{
        HAVE_FFMPEG_AVCODEC_H=y
        pkg_config FFMPEG "libavformat" || return $?
        if check_header "libavcodec/avcodec.h" $FFMPEG_CFLAGS
        then
                HAVE_FFMPEG_AVCODEC_H=n
        else
                check_header "ffmpeg/avcodec.h" $FFMPEG_CFLAGS || return $?
        fi
        # ffmpeg api changes so frequently that it is best to compile the module
        libs="$LDDLFLAGS $FFMPEG_LIBS"
        cflags="$SOFLAGS $FFMPEG_CFLAGS"
        if test "$HAVE_FFMPEG_AVCODEC_H" = y
        then
                cflags="$cflags -DHAVE_FFMPEG_AVCODEC_H"
        fi
        topdir=`dirname "$0"`
        ffmpeg_code=`cat "$topdir"/ffmpeg.c | sed 's/\\\n//g'`
        msg_checking "for successful build of ffmpeg.c"
        if try_compile_link "$ffmpeg_code" $cflags $libs
        then
                msg_result yes
                return 0
        fi
        msg_result no
        return 1
}
```

**Comment by marosd** [comment]
When I run it I have got this in terminal window:

checking for FFMPEG_LIBS (pkg-config)... -pthread -L/home/maros/TILINUX/BUILDCODES/CMUSARM/lib -lavformat -lavcodec -lasound -lbz2 -lz -lswresample -lavutil -lrt -lm  
checking for FFMPEG_CFLAGS (pkg-config)... -I/home/maros/TILINUX/BUILDCODES/CMUSARM/include  
checking for header <libavcodec/avcodec.h>... yes
checking for successful build of ffmpeg.c... no

So I think that it works because it finds libraries and other things... or how else could I try it?

**Comment by marosd** [comment]
I looked into configure and checks.sh scripts... You told me that ffmpeg.c cannot be builded and it makes sense. So I used -c option for compilation (without linking). But if you look into script in configure file there is "try_compile_link". This function is defined in checks.sh script and in this file is:
__src=`tmp_file prog.c`
		__exe=`tmp_file prog`
		echo "$1" > $__src || exit 1
		shift
		__cmd="$CC $__src -o $__exe $CFLAGS $LDFLAGS $@"

As you can see - there is -o option -> that means it is trying to compile and LINK ffmpeg file. ffmpeg.c doesnt have any main() function so it cannot be build as executable file..?

**Comment by flyingmutant** [comment]
@marosd, please use [Markdown](https://guides.github.com/features/mastering-markdown/) to, well, mark down your posts a bit — right now I have to edit each of them manually to be able to read them comfortably.

**Comment by marosd** [comment]
OK, thank you and sorry. Do you any solution of my problem? What is wrong about libswresample. I compile ffmpeg library for ARM without any problem with --enable-shared setting. Cmus ./configure script finds locations -I and -L for FFMPEG, checking for successful build of ffmpeg.c is YES. But after make - as I said - it gives me the error above.

Do you think that problem could be that I have installed FFMPEG into Ubuntu? I mean that there is some conflict between them...hmmm... but it cannot be because -I and -L files are detected OK..

**Comment by marosd** [comment]
Nothing - but ffmpeg is not included in CMus plugins on my ARM device - I cannot play .wma files for example. When I run ./cmus --p on the ARM I don see any ffmpeg plugin.... 

BUT in /lib/cmus/ip/ffmpeg.so file is included!!

**Comment by nefthy** [comment]
Please enable debug (`DEBUG=2 ./configure`), recompile and post last lines few lines of the file ~/cmus-debug.txt after running `cmus --p`

**Comment by marosd** [comment]
hi, i tried it. I had to change DEBUG=2 inside configure script, because DEBUG=2 ./configure didnt work... nevermind. After that cmus-debug.txt file was generated:
'''
ip_load_plugins: /home/maros/TILINUX/BUILDCODES/CMUSA/lib/cmus/ip/mad.so: libmad.so.0: cannot open shared object file: No such file or directory
ip_load_plugins: /home/maros/TILINUX/BUILDCODES/CMUSA/lib/cmus/ip/ffmpeg.so: libavformat.so.56: cannot open shared object file: No such file or directory
ip_load_plugins: /home/maros/TILINUX/BUILDCODES/CMUSA/lib/cmus/ip/cue.so: libcue.so.1: cannot open shared object file: No such file or directory
'''

**Comment by nefthy** [comment]
that probably means there is something wrong with the .so file. Wrong architecture or 64 vs. 32 bit. Wat does `file  /home/maros/TILINUX/BUILDCODES/CMUSA/lib/cmus/ip/ffmpeg.so` say? Compare that to `/lib/libc.so.*`

**Comment by marosd** [comment]
and how could I compare it? Sorry, I dont understand.

My ARM is 32bit version. Maybe it is problem, that I have installed ffmpeg using apt-get before (because I needed it). And now there is conflict between them. Another strange thing is that ./configure file generate wrong CFLAGS and LIBS path in config.mk for LIBMAD. It generates:

MAD_CFLAGS = -I/home/maros/TILINUX/BUILDCODES/CMUSARM/include  
MAD_LIBS = -L/home/maros/TILINUX/BUILDCODES/CMUSARM/lib -lmad 

It points to an old CMUSARM folder that I deleted month ago. Later, I created a new CMUSA folder where I install compiled source codes. For other libraries like FFMPEG, FLAC, VORBIS ./configure script generates good CFLAGS and LIBS path. Only for LIBMAD the path is wrong. I have to edit config.mk file manually and overwrite CMUSARM to CMUSA and after that it works.


I think that I will reinstall UBUNTU and I will try it again (I use UBUNTU only for embedded platforms)... or if you have better idea, I will appreciate it :)
